File: driftwood-e2e/src/authentication/authenticationSteps.js
const {Then, When} = require('cucumber');
const expect = require('unexpected');
const HomePage = require('../pages/HomePage');

When('I authenticate with {string}', async function (provider) {
    const homePage = this._browser.getPageModel(HomePage);
    const header = await homePage.getHeader();
    const loginMenu = await header.getLoginMenu();

    await loginMenu.login(provider);

    profileMenu = await header.getProfileMenu();
});

When('I log out', async function () {
    const homePage = this._browser.getPageModel(HomePage);
    const header = await homePage.getHeader();
    const profileMenu = await header.getProfileMenu();

    await profileMenu.logout();
});

Then('I am not logged in', async function () {
    const homePage = this._browser.getPageModel(HomePage);
    const header = await homePage.getHeader();
    await header.getLoginMenu();
});

Then('the authentication options are:', async function (options) {
    const homePage = this._browser.getPageModel(HomePage);
    const header = await homePage.getHeader();
    const loginMenu = await header.getLoginMenu();

    const providers = await loginMenu.getProviders();

    expect(providers, 'to equal', options.raw()[0]);
});

Then('I am logged in as {string}', async function (expectedUsername) {
    const homePage = this._browser.getPageModel(HomePage);
    const header = await homePage.getHeader();
    const profileMenu = await header.getProfileMenu();

    const userName = await profileMenu.getUserName();

    expect(userName, 'to equal', expectedUsername);
});

File: driftwood-e2e/src/main/kotlin/uk/co/grahamcox/driftwood/e2e/authentication/AuthenticationSteps.kt
package uk.co.grahamcox.driftwood.e2e.authentication

import cucumber.api.java8.En
import io.cucumber.datatable.DataTable
import org.junit.Assert
import org.springframework.beans.factory.annotation.Autowired
import uk.co.grahamcox.driftwood.e2e.browser.Browser
import uk.co.grahamcox.driftwood.e2e.pages.BasePage
import uk.co.grahamcox.driftwood.e2e.pages.HomePage

/**
 * Cucumber steps for handling authentication
 */
class AuthenticationSteps : En {
    /** The browser */
    @Autowired
    private lateinit var browser: Browser

    init {
        When("I authenticate with {string}") { provider: String ->
            BasePage(browser).header.loginMenu.authenticate(provider)
        }

        Then("I am not logged in") {
            Assert.assertFalse(BasePage(browser).header.loggedIn)
        }

        Then("the authentication options are:") { dataTable: DataTable ->
            Assert.assertEquals(dataTable.asList(), BasePage(browser).header.loginMenu.providers)
        }

        Then("I am logged in as {string}") { userName: String ->
            Assert.assertEquals(userName, BasePage(browser).header.profileMenu.userName)
        }

    }
}
